packages/stack/proxy/src/proxy.js

```diff
@@ -121,25 +126,40 @@ export class Proxy {
         const error = modifiedResp.respData.error.message || modifiedResp.respData.error;
         this.logger.error(__(`Error returned from the node: ${error}`));
         const rpcErrorObj = { "jsonrpc": "2.0", "error": { "code": -32603, "message": error }, "id": modifiedResp.respData.id };
-        return this.respondError(transport, rpcErrorObj, isWs);
+        return this.respondError(transport, rpcErrorObj);
       }
-      this.respondOK(transport, modifiedResp.respData, isWs);
+      this.respondOK(transport, modifiedResp.respData);
     }
     catch (resError) {
       // if was an error in response actions (resError), send the error in the response
       const error = resError.message || resError;
       this.logger.error(__(`Error executing response actions: ${error}`));
       const rpcErrorObj = { "jsonrpc": "2.0", "error": { "code": -32603, "message": error }, "id": modifiedRequest.reqData.id };
-      return this.respondError(transport, rpcErrorObj, isWs);
+      return this.respondError(transport, rpcErrorObj);
     }
   }
 
-  forwardRequestToNode(reqData) {
+  forwardRequestToNode(reqData, transport) {
     return new Promise((resolve, reject) => {
-      this.requestManager.send(reqData, (fwdReqErr, result) => {
+      const requestManager = new Web3RequestManager.Manager(this.endpoint);
+      // Watch from subscription data for events
+      if (this.isWs) {
+        requestManager.provider.on('data', (result, deprecatedResult) => {
+          result = result || deprecatedResult;
+          // Just to be clear, while this is *most likely* subscription data, there is no
+          // guarantee it is for subscriptions. This data could contain any data from the
+          // node, which will be forwarded on to the originating socket client connection.
+          this.logger.debug(`Subscription data received from node and forwarded to originating socket client connection: ${JSON.stringify(result)}`);
+          this.respondWs(transport, result);
+        });
+      }
+      requestManager.send(reqData, (fwdReqErr, result) => {
```

Comment:
I might be missing something here, but it seems the primary job of `forwardRequestToNode()` is to take an incoming request from the client, which can be of either transport, and then sends to the node. It then takes the response from the node and sends it back to the client. 

What I'm seeing here is that, when transport is websocket, we start subscribing to any socket data that the request manager receives from the node. If it gets data, it'll do `respondWs()` which by itself will `ws.send()` any response it got. 

Then, in addition to that we have this `requestManager.send()` call which will actually reject/resolve a promise, which is also what the caller of this function seems to expect in https://github.com/embark-framework/embark/pull/1983/files#diff-8e1154cd73e0897f110e3253c0f9477fR111 (notice that we `await forwardRequestToNode()`.

So we have a function that is expected to either resolve or reject (which are one-off actions), but it actually maintains a long-living connection that does neither resolve nor reject.

In addition, it seems that, for every request that goes to the node, we possibly create a new websocket connection via request manager. I'm no expert on websockets, but wouldn't it be enough for the proxy to maintain just a single connection for all possible WS requests that need to go to the node?

I think, given that we're dealing with two different underlying protocols here, that are either one-off calls, or long-living connections, we should maybe have two different APIs that forward requests to the node?

This would also enable reusing a single WS connection from here instead of recreating one for every request that is a WS request.

Reply:
> wouldn't it be enough for the proxy to maintain just a single connection for all possible WS requests that need to go to the node?
So, of course this would be better, but assuming we have only one long-living connection to the node, how will we know which client connection to return the data to? This is the crux of the problem, and one that isn't easy to solve. I can think of two options:
1. Keep one `RequestManager` connection for all requests, **except** when there are requests to `eth_subscribe` (or the parity flavour of this RPC method). In the case of `eth_subscribe`, create a new long-living socket connection (`RequestManager` instance).
2. Try to figure out a way to map `data` from the node to the originating socket connection request. This could possibly involve sending additional data (in the `params` maybe?) to the node, that is hopefully returned in the `data` that can be used for matching connections.

Thoughts?

Reply:
We have gone with Option 1.